### PR TITLE
fix(ga4): added validation and extended support for context.traits for user_properties

### PR DESCRIPTION
**Pull request by mihir-4116**
## Description of the change

- This pull request enhances the support for user_properties within the context.traits feature. Based on customer feedback and recommendations from the customer success team, we have made changes to automatically consider the properties inside context.traits as user_properties.

- To achieve this, we have added support for context.traits while maintaining the existing mapping properties.user_properties. This ensures compatibility with the previous implementation while expanding the capabilities of user_properties.

- To align with the requirements of GA4 (Google Analytics 4), we have made modifications to drop any properties that are not of type string or number or boolean. GA4 silently drops events if any user_property is an array or object, so we are handling this scenario proactively.

- Additionally, we have transformed the user_properties to conform to the structure expected by the measurement protocol. This ensures seamless integration and accurate reporting within the GA4 framework.

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

**Review by koladilip — COMMENTED** [review]
why is this change?

**Comment by sanpj2292** [comment]
@mihir-4116 can we add some description to have context about why we're making this change & also have a description about what is the change exactly ?

**Review by koladilip — DISMISSED** [review]

File: src/v0/destinations/ga4/transform.js [hunk 1]
```diff
@@ -262,12 +262,22 @@ const responseBuilder = (message, { Config }) => {
...
+    const validatedUserProperties = {};
+
+    for (const key of userPropertiesKeys) {
+      const value = userProperties[key];
+      if (typeof value === 'string' || typeof value === 'number') {
+        validatedUserProperties[key] = { value };
+      }
+    }
+
+    rawPayload.user_properties = validatedUserProperties;
```
Comment (utsabc) [inline src/v0/destinations/ga4/transform.js:280]: Even we are also dropping silently any properties which are not `string` or `number`

Comment (mihir-4116) [reply src/v0/destinations/ga4/transform.js:280]: yes, if we don't drop it then GA4 will silently drop whole event

Comment (utsabc) [reply src/v0/destinations/ga4/transform.js:280]: Lets remove this validation as we discussed with AMy we will not do this now

**Review by utsabc — APPROVED** [review]

**Review by utsabc — DISMISSED** [review]

**Comment by krishna2020** [comment]
event name validation is not part of this PR, can the name be updated ?

**Review by krishna2020 — DISMISSED** [review]

**Review by utsabc — APPROVED** [review]

**Review by utsabc — DISMISSED** [review]

File: src/v0/destinations/ga4/utils.js [hunk 2]
```diff
@@ -336,30 +313,85 @@ const validateEventName = (event) => {
...
+  const isValidValue = isDefinedAndNotNull(value) && isValidValueType(value);
+
+  return isValidKey && isValidValue;
+};
+
+/**
+ * Function to validate and prepare user_properties
+ * @param {*} message
+ */
+const prepareUserProperties = (message) => {
```
Comment (krishna2020) [inline src/v0/destinations/ga4/utils.js:364]: Good refactoring, now write unit test for `prepareUserProperties` in file `util.test.js` --> which makes easy & convenient to cover varied test cases

Comment (krishna2020) [reply src/v0/destinations/ga4/utils.js:364]: you can focus on test case payloads only with combination of 2 fields 'properties.user_properties', 'context.traits'

**Review by utsabc — DISMISSED** [review]

**Review by krishna2020 — APPROVED** [review]

**Review by utsabc — APPROVED** [review]

**Pull request merged**